### DISCUSSION TITLE
How to avoid `Nokogiri::HTML.parse` behavior

### DISCUSSION THREAD

**Discussion by YusukeSuzuki** [sparklemotion/nokogiri#2998] [General Q&A] [answered]
`Nokogiri::HTML.parse('<div>text</div>').to_html` creates 

```
"<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.0 Transitional//EN\" \"http://www.w3.org/TR/REC-html40/loose.dtd\">\n<html><body><p>test</p></body></html>\n"
```

but `Nokogiri::HTML.parse('text').to_html` creates

```
"<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.0 Transitional//EN\" \"http://www.w3.org/TR/REC-html40/loose.dtd\">\n<html><body><p>test</p></body></html>\n"
```

How do I work around this bug? without `Nokogiri::HTML::DocumentFragment.parse`

**Comment by flavorjones** [comment, marked as answer]
@YusukeSuzuki Thanks for asking this question. What you're seeing is how libxml2 (the underlying HTML4 parser) constructs a document around this fragment.

I'm curious why you don't want to use `DocumentFragment` as this is exactly the use case it addresses. Neither `<div>text</div>` nor `text` is a Document.

You may also want to use `Nokogiri::HTML5` which uses libgumbo instead of libxml2, and that library follows the precise rules in the HTML5 spec around document structure:

```ruby
Nokogiri::HTML5.parse('<div>text</div>').to_html
# => "<html><head></head><body><div>text</div></body></html>"

Nokogiri::HTML5.parse('text').to_html
# => "<html><head></head><body>text</body></html>"
```

But really, again, I suggest you consider using fragment parsing:

```ruby
Nokogiri::HTML5.fragment('<div>text</div>').to_html
# => "<div>text</div>"

Nokogiri::HTML5.fragment('text').to_html
# => "text"
```

**Reply by flavorjones** [reply]
By the way, in your original post I get different results than is written, which I assume is a copy-paste error? You report you see `<p>`:

```ruby
Nokogiri::HTML.parse('<div>text</div>').to_html
# => "<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.0 Transitional//EN\" \"http://www.w3.org/TR/REC-html40/loose.dtd\">\n<html><body><p>test</p></body></html>\n"
```

but I see `<div>`:

```ruby
Nokogiri::HTML4.parse('<div>text</div>').to_html
# => "<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.0 Transitional//EN\" \"http://www.w3.org/TR/REC-html40/loose.dtd\">\n" +
#    "<html><body><div>text</div></body></html>\n"
```

Just wanted to make sure I understand what you're asking.
